Make createEndpointService() and deleteEndpointService() plugin interface methods.

humblec opened the pull request:
Why this change?

In some setups, after creation of dynamic PVs and before mounting/using these  PVs in a pod, the endpoint/service got mistakenly deleted by the user/developer. By making these methods 'plugin' specific, we can call it from mounter if there are scenarios where the endpoint and service got wiped in between accidentally.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45528)
<!-- Reviewable:end -->

humblec commented:
/release-note-none

humblec commented:
/assign @jsafrane @rootfs

humblec commented:
@k8s-bot ok to test

rootfs commented:
@k8s-bot ok to test

humblec commented:
@rootfs you are correct, these are provisioner methods. The main reason to make this change is due to below.
In some setups, after creation of PVs and before mounting of the PV in a pod, the endpoint got mistakenly deleted by the user. So, what I am planning with this change is calling this function at mount time, if some how the endpoint and service got wiped. I would have mentioned it in the commit message. I will update the  commit message if you ack on this change.

humblec commented:
@rootfs the PR note reflect the reason for this change. Thanks !

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,49 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
 		glog.Errorf("glusterfs: failed to get endpoints %s[%v]", epName, err)
-		return nil, err
+
+		class, err := volutil.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
```
recreate endpoint if:
- spec != nil
- err indicates the endpoint doesn't exist.

humblec replied:
Done.

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,49 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+
+		class, err := volutil.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
+		if err != nil {
+			return nil, fmt.Errorf("glusterfs: failed to get storageclass: %v", err)
```
PV may not be created via storage class. Tune the message a bit.

humblec replied:
Done.

humblec commented:
@rootfs Addressed the comments. Please review.

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,51 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		if spec != nil && spec.PersistentVolume.Annotations["kubernetes.io/createdby"] == "heketi-dynamic-provisioner" {
+			class, err := volutil.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to get storageclass when recreating endpoint/service: %v", err)
```
If the PV has no storage class, don't treat it as an error.

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,51 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+
+			cfg, err := parseClassParameters(class.Parameters, plugin.host.GetKubeClient())
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to parse storageclass parameter %v when recreating endpoint/service", err)
```
failed to parse storage class parameter. error: %v

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,51 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+			claim := spec.PersistentVolume.Spec.ClaimRef.Name
+			_, _, err = plugin.createEndpointService(ns, epName, endpointIPs, claim)
+			if err != nil {
+				glog.Errorf("glusterfs: failed to recreate endpoint/service: %v", err)
```
There could be a race condition when two PVs are used simultaneously and yet one PV creates the endpoint before the other.

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,51 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+				glog.Errorf("glusterfs: failed to recreate endpoint/service: %v", err)
+				return nil, fmt.Errorf("failed to recreate endpoint/service %v", err)
+			}
+			glog.V(1).Infof("glusterfs: endpoint/service [%v] successfully recreated ", epName)
```
V(3)

humblec commented:
@rootfs Done.. Thanks !!!

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		if spec != nil && spec.PersistentVolume.Annotations["kubernetes.io/createdby"] == "heketi-dynamic-provisioner" {
+			class, err := volutil.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to recreate endpoint/service, error: %v", err)
```
failed to get storage class

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
-		return nil, err
+	if err != nil && errors.IsNotFound(err) {
+		glog.Errorf("glusterfs: failed to get endpoint %s[%v]", epName, err)
+		if spec != nil && spec.PersistentVolume.Annotations["kubernetes.io/createdby"] == "heketi-dynamic-provisioner" {
```
make heketi-dynamic-provisioner a const and replace it in provision() too

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+
+			cfg, err := parseClassParameters(class.Parameters, plugin.host.GetKubeClient())
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to recreate endpoint/service, error: %v", err)
```
failed to parse parameters

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+			scConfig := *cfg
+			cli := gcli.NewClient(scConfig.url, scConfig.user, scConfig.secretValue)
+			if cli == nil {
+				return nil, fmt.Errorf("glusterfs: failed to recreate endpoint/service, error: %v", err)
```
failed to create heketi client

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+			volumeID := dstrings.TrimPrefix(source.Path, volPrefix)
+			volInfo, err := cli.VolumeInfo(volumeID)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to recreate endpoint/service, error: %v", err)
```
failed to get volume info

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+
+			endpointIPs, err := getClusterNodes(cli, volInfo.Cluster)
+			if err != nil {
+				return nil, fmt.Errorf("glusterfs: failed to recreate endpoint/service, error: %v", err)
```
failed to get cluster nodes

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +141,52 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
-	if err != nil {
-		glog.Errorf("glusterfs: failed to get endpoints %s[%v]", epName, err)
-		return nil, err
+	if err != nil && errors.IsNotFound(err) {
```
a bit more paranoid but validate that `epName` is created using the same schema that provision uses.

humblec replied:
Done.

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +144,57 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		claim := spec.PersistentVolume.Spec.ClaimRef.Name
+		checkEpName := dynamicEpSvcPrefix + claim
+		if epName != checkEpName {
+			return nil, fmt.Errorf("failed to get proper endpoint name, error: %v", err)
```
`glusterfs: failed to get endpoints %s, error %v` since in this case, failed validation indicates the volume is not provisioned by provision().

humblec replied:
Done.

humblec commented:
@k8s-bot verify test this

rootfs commented:
/approve
@k8s-bot verify test this

@humblec squash commits, then lgtm

humblec committed:
Recreate endpoint/service if it lost in between.

In some setups, after creation of dynamic PVs and before mounting/using
these PVs in a pod, the endpoint/service got mistakenly deleted by the
user/developer. By making these methods 'plugin' specific, we can call
it from mounter if there are scenarios where the endpoint and service
got wiped in between accidentally.

Signed-off-by: Humble Chirammal hchiramm@redhat.com

rootfs commented:
/lgtm

humblec commented:
Thanks a lot @rootfs !

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +144,57 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		}
+		glog.Errorf("glusterfs: failed to get endpoint %s[%v]", epName, err)
+		if spec != nil && spec.PersistentVolume.Annotations["kubernetes.io/createdby"] == heketiAnn {
+			class, err := volutil.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
```
nodes cannot look up arbitrary secrets based on values in storage class parameters...

liggitt replied:
I also thought only the create/delete, attach/detach phases were supposed to resolve storage class info... isn't mount only supposed to use info in the PV?

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -141,12 +144,57 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+
+			// Give an attempt to recreate endpoint/service.
+
+			_, _, err = plugin.createEndpointService(ns, epName, endpointIPs, claim)
```
nodes do not have permission to create endpoints or services... this will fail on any cluster running RBAC
